main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunAndClosureClasses.scala

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright 2021 Jonathan Lindegaard Starup
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.jvm
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.ReducedAst.{Def, Root}
+import ca.uwaterloo.flix.language.ast.{MonoType, Symbol}
+import ca.uwaterloo.flix.language.phase.jvm.BytecodeInstructions.InstructionSet
+import ca.uwaterloo.flix.language.phase.jvm.JvmName.MethodDescriptor
+import ca.uwaterloo.flix.util.ParOps
+import org.objectweb.asm.Opcodes._
+import org.objectweb.asm.{ClassWriter, Label, MethodVisitor, Opcodes}
+
+/**
+  * Generates byte code for the function and closure classes.
+  */
+object GenFunAndClosureClasses {
+
+  /**
+    * Returns a map of function- and closure-classes for the given set `defs`.
+    */
+  def gen(defs: Map[Symbol.DefnSym, Def])(implicit root: Root, flix: Flix): Map[JvmName, JvmClass] = {
+    ParOps.parAgg(defs.values, Map.empty[JvmName, JvmClass])({
+
+      case (macc, closure) if isClosure(closure) =>
+        val closureType = JvmOps.getClosureClassType(closure.sym)
+        val closureName = closureType.name
+        val code = genCode(closureType, Closure, closure)
+        macc + (closureName -> JvmClass(closureName, code))
+
+      case (macc, defn) if isFunction(defn) =>
+        flix.subtask(defn.sym.toString, sample = true)
+        val functionType = JvmOps.getFunctionDefinitionClassType(defn.sym)
+        val functionname = functionType.name
+        val code = genCode(functionType, Function, defn)
+        macc + (functionname -> JvmClass(functionname, code))
+
+      case (macc, _) =>
+        macc
+    }, _ ++ _)
+  }
+
+  private def isClosure(defn: Def): Boolean = defn.cparams.nonEmpty
+
+  private def isFunction(defn: Def): Boolean = defn.cparams.isEmpty
```

Comment:
Something about excluded middle, could define isClosure in terms of isFunction

Reply:
Yeah but i like it for readability. I can delete one of them if you want